### PR TITLE
[KRYS-3548] Add new transaction setting

### PR DESCRIPTION
**Pull request by MrQKyber**

### CODE REVIEW

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+}
+
+extension TransactionSettingsViewController: UITableViewDataSource {
+  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+    return 4
+  }
+  
+  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+    switch indexPath.row {
+    case 0:
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:269]: Should not switch indexPath like this. 
You should create an enum for row types 
```
enum TxSettingRowType {
    case slippage 
    case expertMode
    case basicMode
    case advanceMode
    ...
}
```
And create a list of rows: 
```
let rows = [.slippage, .advancedMode, .expertMode]
```
Then
```
switch rows[indexPath.row] {
   case .slippage:
        /// dequeue slippage cell
    case .expertMode:
        /// dequeue expert mode cell 
    .....
}
````

With this solution, all case are straight and no need to if else like case 2: 

Using case 0, 1, 2, ... will make some crash issues when adding/removing rows without updating this switch block.

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+          cell.fillFormValues()
+          cell.updateUI()
+          return cell
+        } else {
+          let cell = tableView.dequeueReusableCell(
+            withIdentifier: SettingBasicModeCell.cellID,
+            for: indexPath
+          ) as! SettingBasicModeCell
+          cell.cellModel = viewModel.basicModeCellModel
+          cell.updateUI()
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:311]: Why not `cell.updateUI(with viewModel: SettingBasicModeCell)`

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+          return cell
+        }
+      }
+    case 3:
+      let cell = tableView.dequeueReusableCell(
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:316]: Use `let cell = tableView.dequeueReusableCell(SettingExpertModeSwitchCell.self, indexPath: indexPath)!` and no need casting. 
Same to other cells.

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    slippageCellModel.resetData()
+    basicModeCellModel.resetData()
+    basicAdvancedCellModel.resetData()
+    advancedModeCellModel.resetData()
+    segmentedCellModel.resetData()
+    switchExpertMode.resetData()
+  }
+  
+  func update(priorityFee: String?, maxGas: String?, gasLimit: String?, nonceString: String?) {
+    if let notNil = priorityFee {
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:160]: Please don't use variable names like `notNil`, `unwraped`, let use
```
if let priorityFee = priorityFee {

}
```

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+      advancedModeCellModel.maxPriorityFeeString = notNil
+      self.basicModeCellModel.selectedIndex = -1
+    }
+    
+    if let notNil = maxGas {
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:165]: Same as above

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+      basicAdvancedCellModel.gasPriceString = notNil
+      advancedModeCellModel.maxFeeString = notNil
+      self.basicModeCellModel.selectedIndex = -1
+    }
+    
+    if let notNil = gasLimit {
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:171]: Same as above

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+      advancedModeCellModel.gasLimitString = notNil
+      basicAdvancedCellModel.gasLimitString = notNil
+      self.basicModeCellModel.selectedIndex = -1
+    }
+    
+    if let notNil = nonceString, let nonceInt = Int(notNil) {
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:177]: Same as above

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+//  Created by Ta Minh Quan on 08/08/2022.
+//
+
+import UIKit
+import BigInt
+
+typealias BasicSettingsInfo = (type: KNSelectedGasPriceType, value: BigInt)
+typealias AdvancedSettingsInfo = (maxPriority: String, maxFee: String, gasLimit: String)
+
+class TransactionSettingsViewModel {
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:14]: Please move ViewModel to another file. This file is too large

File: KyberNetwork/KyberNetwork/Screens/CommonViews/Popups/GasFeeSelector/GasFeeSelectorPopupViewController.swift [hunk 9]
```diff
@@ -11,7 +11,7 @@ import APIKit
 import JSONRPCKit
 
 protocol GasFeeSelectorPopupViewControllerDelegate: class {
-  func gasFeeSelectorPopupViewController(_ controller: GasFeeSelectorPopupViewController, run event: GasFeeSelectorPopupViewEvent)
+  func gasFeeSelectorPopupViewController(_ controller: KNBaseViewController, run event: GasFeeSelectorPopupViewEvent)
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/Popups/GasFeeSelector/GasFeeSelectorPopupViewController.swift:14]: Should create a `BaseGasFeeSelectorViewController`

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/Cells/SettingAdvancedModeFormCell.swift [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  
+  func resetData() {
+    gasLimit = gasLimitInit
+    selectedType = .medium
+  }
+}
+
+class SettingAdvancedModeFormCell: UITableViewCell {
+
+  static let cellID: String = "SettingAdvancedModeFormCell"
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/Cells/SettingAdvancedModeFormCell.swift:64]: No need to create this cellID

**Comment by tungnguyen20** [comment]
Expert mode info button not clickable

**Comment by tungnguyen20** [comment]
Screen Shot 2022-08-12 at 17 39 04

1. Expert Mode text color must be orange 
2. Confirm text field placeholder 
3. Confirm text field background color 
4. Cancel button title color (must be white)

**Comment by tungnguyen20** [comment]
Expert mode switch must return to off if user click Cancel

**Comment by tungnguyen20** [comment]
Please get nonce inside the Popup, not call from the parent ViewController. You can refer to line 839 in the `GasFeeSelectorPopupViewController` class.

**Comment by tungnguyen20** [comment]
Field titles is not underlined

**Comment by tungnguyen20** [comment]
Input fields keyboard type must be numberPad

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+//
+//  Created by Ta Minh Quan on 08/08/2022.
+//
+
+import UIKit
+import BigInt
+import Result
+
+typealias BasicSettingsInfo = (type: KNSelectedGasPriceType, value: BigInt)
+typealias AdvancedSettingsInfo = (maxPriority: String, maxFee: String, gasLimit: String)
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:13]: Please create a struct instead

Comment (tungnguyen20) [reply KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:13]: Or please use `SwapTransactionSettings`

Comment (MrQKyber) [reply KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/TransactionSettingsViewController.swift:13]: I make changes follow your request. Plz help to implement handle for output in block
Screen Shot 2022-08-17 at 13 45 10
k

File: KyberNetwork/Resources/Strings.swift [hunk 12]
```diff
@@ -206,6 +206,7 @@ struct Strings {
   static let swapTxnFeeInfo = "swap.txn_fee_i".toBeLocalised()
   static let swapTxnMaxFeeInfo = "swap.txn_fee_max_i".toBeLocalised()
   static let swapPriceImpactInfo = "swap.price_impact_i".toBeLocalised()
+  static let advancedModeWarningText = "Expert Mode turns off the ‘Confirm’ transaction prompt and alows high slippage trades that can result in bad rates and lost funds.\n\nPlease type the word ‘confirm’ below to enable Expert Mode."
```
Comment (tungnguyen20) [inline KyberNetwork/Resources/Strings.swift:209]: Move this text to Localizations file

File: KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/Cells/SettingBasicAdvancedFormCell.swift [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+      gasPriceErrorLabel.text = "Max Fee is higher than necessary"
+      gasPriceContainerView.rounded(color: UIColor.Kyber.textRedColor, width: 1, radius: 16)
+    case .none, .empty:
+      gasPriceErrorLabel.text = ""
+      gasPriceContainerView.rounded(color: .clear, width: 0, radius: 16)
+    }
+    
+    switch cellModel.advancedGasLimitErrorStatus {
+    case .low:
+      gasLimitErrorLabel.text = "Gas limit must be at least 21000"
```
Comment (tungnguyen20) [inline KyberNetwork/KyberNetwork/Screens/CommonViews/TransactionSettings/Cells/SettingBasicAdvancedFormCell.swift:147]: Must localize and create a constant for 21000

**Comment by tungnguyen20** [comment]
Anh @MrQKyber please localize strings and move colors to constants

**Review by tungnguyen20 — APPROVED** [review]

**Pull request merged**